KAFKA-18915: Rewrite AdminClientRebootstrapTest to cover the current scenario

clarkwtc opened the pull request:
Follow-up: https://github.com/apache/kafka/pull/19094

We need to rewrite `AdminClientRebootstrapTest` to cover the current scenario.
Added the admin client with rebootstrap disabled, as the admin client uses the default `AdminClientRebootstrap` configuration setting.

Default `AdminClientRebootstrap` config:

> properties.put(CommonClientConfigs.METADATA_RECOVERY_STRATEGY_CONFIG, "rebootstrap");
       properties.put(CommonClientConfigs.METADATA_RECOVERY_REBOOTSTRAP_TRIGGER_MS_CONFIG, "3000000");
       properties.put(CommonClientConfigs.SOCKET_CONNECTION_SETUP_TIMEOUT_MS_CONFIG, "10000");
       properties.put(CommonClientConfigs.SOCKET_CONNECTION_SETUP_TIMEOUT_MAX_MS_CONFIG, "30000");
       properties.put(CommonClientConfigs.RECONNECT_BACKOFF_MS_CONFIG, "50L");
       properties.put(CommonClientConfigs.RECONNECT_BACKOFF_MAX_MS_CONFIG, "1000L");

The test case for the admin client with enabled rebootstrap
<img width="1547" alt="Screenshot 2025-03-11 at 11 57 00 PM" src="https://github.com/user-attachments/assets/f75621f1-9286-477d-85ce-7d7a80881db5" />

The test case for the admin client with disabled rebootstrap
<img width="1555" alt="Screenshot 2025-03-11 at 11 57 38 PM" src="https://github.com/user-attachments/assets/a92049ab-8106-48ef-b4c3-1cdb27ef5920" />

Reviewers: Jhen-Yung Hsu <jhenyunghsu@gmail.com>, TengYao Chi <kitingiao@gmail.com>, Ken Huang <s7133700@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

Commit:
feat: Rewrite AdminClientRebootstrapTest to cover the current scenario

frankvicky commented on core/src/test/java/kafka/test/api/AdminClientRebootstrapTest.java:
```diff
@@ -20,91 +20,85 @@
-            // Only the server 0 is available for the admin client during the bootstrap.
-            TestUtils.waitForCondition(() -> admin.listTopics().names().get(timeout, TimeUnit.MINUTES).contains(topic),
+            // Only the broker 1 is available for the admin client during the bootstrap.
+            TestUtils.waitForCondition(() -> admin.listTopics().names().get(timeout, TimeUnit.MINUTES).contains(TOPIC),
```
Out of curiosity, why did we have a 5-minute timeout here?

clarkwtc replied:
Thank you for your comments.
Maybe it’s a bit too long to wait here.
I updated to 60 seconds.

frankvicky reviewed (commented):
@clarkwtc : Thanks for the patch.

Yunyung commented on core/src/test/java/kafka/test/api/AdminClientRebootstrapTest.java:
```diff
@@ -20,91 +20,85 @@
+        clusterInstance.startBroker(broker0);
+
+        // The broker 1, originally cached during the bootstrap, is offline.
+        // The admin client will throw a TimeoutException because the brokers are offline during the bootstrap list
```
This code comment is not correct.

clarkwtc replied:
Thank you for your comments.
Sure, this situation refers to the brokers being offline since they were cached during the bootstrap.
I updated the code comments.

chia7712 commented:
@clarkwtc cloud you please move this test to clients-integration-tests module?

Yunyung commented on core/src/test/java/kafka/test/api/AdminClientRebootstrapTest.java:
```diff
@@ -20,91 +20,85 @@
+        // The broker 1, originally cached during the bootstrap, is offline.
+        // The admin client will throw a TimeoutException because the brokers are offline during the bootstrap list
+        assertThrows(TimeoutException.class, () -> admin.listTopics().names().get(5, TimeUnit.SECONDS));
+        // The admin client needs to wait the default timeout for other threads because the brokers are offline.
```
Ditto.

Yunyung commented on core/src/test/java/kafka/test/api/AdminClientRebootstrapTest.java:
```diff
@@ -20,91 +20,85 @@
+        // The admin client will throw a TimeoutException because the brokers are offline during the bootstrap list
+        assertThrows(TimeoutException.class, () -> admin.listTopics().names().get(5, TimeUnit.SECONDS));
+        // The admin client needs to wait the default timeout for other threads because the brokers are offline.
+        admin.close(Duration.ofSeconds(0));
```
nit: Duration.ZERO

clarkwtc replied:
I've fixed it.

Commit:
chore: Make it better.

Commit:
Merge branch 'trunk' into KAFKA-18915-rewrite-adminclient_rebootstrap_test-testing-scenario

Commit:
feat: Move this test to clients-integration-tests

clarkwtc commented:
@chia7712 
I already moved this test to clients-integration-tests.

chia7712 commented:
@clarkwtc please check the build error

Commit:
fix: Check style

Commit:
Merge branch 'trunk' into KAFKA-18915-rewrite-adminclient_rebootstrap_test-testing-scenario

clarkwtc commented:
@chia7712 
Sorry, I missed it.
I've fixed it.

chia7712 commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/admin/AdminClientRebootstrapTest.java:
```diff
@@ -0,0 +1,101 @@
+    private static final int PARTITIONS = 2;
+
+    @ClusterTest(
+        brokers = 2,
```
`2` -> `PARTITIONS`

clarkwtc replied:
Thank you for your comments.
I've fixed it.

chia7712 commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/admin/AdminClientRebootstrapTest.java:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    @ClusterTest(
+        brokers = 2,
```
ditto

clarkwtc replied:
I've fixed it.

chia7712 commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/admin/AdminClientRebootstrapTest.java:
```diff
@@ -0,0 +1,101 @@
+        brokers = 2,
+        types = {Type.KRAFT},
+        serverProperties = {
+            @ClusterConfigProperty(key = TopicConfig.UNCLEAN_LEADER_ELECTION_ENABLE_CONFIG, value = "true"),
```
we don't need those configs, since this test case does not use either producer or consumer.

clarkwtc replied:
Sorry, I wrote the config for ProducerClientRebootstrapTest, but I messed up how to use it.

chia7712 commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/admin/AdminClientRebootstrapTest.java:
```diff
@@ -0,0 +1,101 @@
+        brokers = 2,
+        types = {Type.KRAFT},
+        serverProperties = {
+            @ClusterConfigProperty(key = TopicConfig.UNCLEAN_LEADER_ELECTION_ENABLE_CONFIG, value = "true"),
```
ditto

clarkwtc replied:
I've fixed it.

chia7712 commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/admin/AdminClientRebootstrapTest.java:
```diff
@@ -0,0 +1,101 @@
+
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class AdminClientRebootstrapTest {
```
maybe we can rename it to `ClientRebootstrapTest` and merge producer/consumer into this class

clarkwtc replied:
No problem; 
I also changed `ClientRebootstrapTest` file location from `org/apache/kafka/clients/admin` to `org/apache/kafka/clients` and renamed functions.

chia7712 reviewed (commented):
@clarkwtc thanks for your patch.

m1a2st commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/admin/AdminClientRebootstrapTest.java:
```diff
@@ -0,0 +1,101 @@
+            admin.createTopics(List.of(new NewTopic(TOPIC, PARTITIONS, (short) 2)));
+
+            // Only the broker 1 is available for the admin client during the bootstrap.
+            Assertions.assertDoesNotThrow(() -> admin.listTopics().names().get(timeout, TimeUnit.SECONDS).contains(TOPIC));
```
nit: static import

clarkwtc replied:
Thank you for your comments.
I've fixed it.

m1a2st reviewed (commented):
Thanks for this patch, a little comment

Commit:
chore: Make it better

- Change Assertions to static import
- Redefine AdminClientRebootstrapTest to ClientRebootstrapTest
- Remove unused config UNCLEAN_LEADER_ELECTION_ENABLE_CONFIG

Commit:
Merge branch 'trunk' into KAFKA-18915-rewrite-adminclient_rebootstrap_test-testing-scenario

chia7712 commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/ClientRebootstrapTest.java:
```diff
@@ -0,0 +1,97 @@
+
+public class ClientRebootstrapTest {
+    private static final String TOPIC = "topic";
+    private static final int PARTITIONS = 2;
```
the purpose is to ensure each broker has a replica, and hence it should be renamed to "REPLICAS"

clarkwtc replied:
I've fixed it.

chia7712 commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/ClientRebootstrapTest.java:
```diff
@@ -0,0 +1,97 @@
+        clusterInstance.shutdownBroker(broker0);
+
+        try (var admin = clusterInstance.admin()) {
+            admin.createTopics(List.of(new NewTopic(TOPIC, PARTITIONS, (short) 2)));
```
All we need is "one partition" and "two replicas", right?

clarkwtc replied:
Sure, we only need one partition across two brokers because the topic will be replicated to both brokers.

chia7712 reviewed (commented):
@clarkwtc thanks for this patch.

Commit:
fix: Redefine test case to use only one partitions.

Commit:
Merge branch 'trunk' into KAFKA-18915-rewrite-adminclient_rebootstrap_test-testing-scenario

chia7712 commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/ClientRebootstrapTest.java:
```diff
@@ -0,0 +1,97 @@
+            admin.createTopics(List.of(new NewTopic(TOPIC, 1, (short) REPLICAS)));
+
+            // Only the broker 1 is available for the admin client during the bootstrap.
+            assertDoesNotThrow(() -> admin.listTopics().names().get(timeout, TimeUnit.SECONDS).contains(TOPIC));
```
We should verify the existence of `TOPIC` here ...

chia7712 commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/ClientRebootstrapTest.java:
```diff
@@ -0,0 +1,97 @@
+            // The broker 1, originally cached during the bootstrap, is offline.
+            // However, the broker 0 from the bootstrap list is online.
+            // Should be able to list topics again.
+            assertDoesNotThrow(() -> admin.listTopics().names().get(timeout, TimeUnit.SECONDS).contains(TOPIC));
```
ditto

chia7712 commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/ClientRebootstrapTest.java:
```diff
@@ -0,0 +1,97 @@
+        admin.createTopics(List.of(new NewTopic(TOPIC, 1, (short) REPLICAS)));
+
+        // Only the broker 1 is available for the admin client during the bootstrap.
+        assertDoesNotThrow(() -> admin.listTopics().names().get(60, TimeUnit.SECONDS).contains(TOPIC));
```
ditto

chia7712 commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/ClientRebootstrapTest.java:
```diff
@@ -0,0 +1,97 @@
+
+        clusterInstance.shutdownBroker(broker0);
+
+        var admin = clusterInstance.admin(Map.of(CommonClientConfigs.METADATA_RECOVERY_STRATEGY_CONFIG, "none"));
```
We should use try-finally​ to make sure the admin gets closed even if an assertion fails!
